File: ai-service/app/rag/embeddings.py
"""Embedding provider abstraction (PRD dim 1536). No live calls in tests."""

from __future__ import annotations

import hashlib
import math
from typing import Protocol

from app.core.config import get_settings
from app.core.logging import get_logger

logger = get_logger("ai-embeddings")


class EmbeddingClient(Protocol):
    name: str
    dim: int

    def embed(self, texts: list[str]) -> list[list[float]]: ...


def deterministic_embedding(text: str, dim: int) -> list[float]:
    vec = [0.0] * dim
    for token in text.lower().split():
        token = "".join(c for c in token if c.isalnum())
        if not token:
            continue
        digest = hashlib.sha256(token.encode()).digest()
        vec[int.from_bytes(digest[0:2], "big") % dim] += 1.0
        vec[int.from_bytes(digest[2:4], "big") % dim] += 0.5
    norm = math.sqrt(sum(v * v for v in vec))
    if norm > 0:
        vec = [v / norm for v in vec]
    return vec


class MockEmbeddingClient:
    name = "mock"

    def __init__(self, dim: int = 1536):
        self.dim = dim

    def embed(self, texts: list[str]) -> list[list[float]]:
        return [deterministic_embedding(t, self.dim) for t in texts]


class AnthropicEmbeddingClient:
    """Real embedding provider (Voyage-compatible HTTP) with safe fallback.

    - Key only from environment (EMBEDDING_API_KEY / LLM_API_KEY fallback).
    - Configurable model via EMBEDDING_MODEL (default voyage-3-lite).
    - Timeout + one safe retry (embeddings are idempotent).
    - Dimension validated before returning; mismatches raise (callers must
      not insert wrong-dim vectors into pgvector).
    - Tests set no key and inject MockEmbeddingClient, so no live calls occur.
    """

    name = "anthropic"

    def __init__(self, dim: int = 1536):
        self.dim = dim

    def embed(self, texts: list[str]) -> list[list[float]]:
        import json
        import time
        import urllib.error
        import urllib.request

        from app.core.config import get_settings as _settings

        settings = _settings()
        api_key = settings.embedding_api_key or ""
        if not api_key:
            logger.warning("provider embeddings not wired (no key); deterministic fallback (tests/dev)")
            return [deterministic_embedding(t, self.dim) for t in texts]
        base = (settings.embedding_base_url or "https://api.voyageai.com").rstrip("/")
        expected = self.dim or settings.embedding_dim or 1536
        last_error: Exception | None = None
        for attempt in range(2):
            started = time.monotonic()
            try:
                payload = {"model": settings.embedding_model or "voyage-3-lite", "input": texts}
                req = urllib.request.Request(
                    base + "/v1/embeddings",
                    data=json.dumps(payload).encode(),
                    headers={"content-type": "application/json", "authorization": f"Bearer {api_key}"},
                    method="POST",
                )
                with urllib.request.urlopen(req, timeout=30) as res:
                    body = json.loads(res.read().decode())
                latency_ms = int((time.monotonic() - started) * 1000)
                vectors = [d.get("embedding", []) for d in body.get("data", [])]
                if len(vectors) != len(texts):
                    raise ValueError("Embedding provider returned mismatched count")
                for v in vectors:
                    validate_dim(v, expected)
                logger.info("embeddings completed provider=voyage count=%d latency_ms=%d", len(texts), latency_ms)
                return [[float(x) for x in v] for v in vectors]
            except urllib.error.HTTPError as exc:
                logger.warning("embedding provider http error status=%s attempt=%d", exc.code, attempt)
                last_error = Exception(f"Embedding provider error: {exc.code}")
                continue
            except Exception as exc:
                logger.warning("embedding provider failed attempt=%d err=%s", attempt, type(exc).__name__)
                last_error = exc if isinstance(exc, Exception) else Exception(str(exc))
                continue
        # Fail closed on provider errors in production, but never break tests:
        # tests inject mocks / set no key (handled above). When a key IS set
        # and the provider fails, surface the error instead of silently
        # indexing wrong vectors.
        raise last_error or Exception("Embedding provider failed")


def validate_dim(vec: list[float], expected: int) -> None:
    if not isinstance(vec, list) or len(vec) != expected:
        raise ValueError(f"Embedding dimension mismatch: got {len(vec) if isinstance(vec, list) else 'invalid'}, expected {expected}")
    if not all(isinstance(v, (int, float)) and v == v and abs(v) != float("inf") for v in vec):
        raise ValueError("Embedding contains non-finite values")


_override: EmbeddingClient | None = None


def set_embedding_client(client: EmbeddingClient | None) -> None:
    global _override
    _override = client


def get_embedding_client() -> EmbeddingClient:
    if _override is not None:
        return _override
    dim = get_settings().embedding_dim or 1536
    if get_settings().llm_provider.lower() == "mock":
        return MockEmbeddingClient(dim)
    return AnthropicEmbeddingClient(dim)


def cosine(a: list[float], b: list[float]) -> float:
    n = min(len(a), len(b))
    dot = sum(a[i] * b[i] for i in range(n))
    na = math.sqrt(sum(v * v for v in a[:n]))
    nb = math.sqrt(sum(v * v for v in b[:n]))
    if na == 0 or nb == 0:
        return 0.0
    return dot / (na * nb)

File: backend/src/ai/embeddings.ts
import { createHash } from 'node:crypto';
import { getConfig } from '../config';
import { logger } from '../logger';

export const EMBEDDING_DIM = 1536;

export interface EmbeddingProvider {
  name: string;
  dim: number;
  embed(texts: string[]): Promise<number[][]>;
}

let overrideProvider: EmbeddingProvider | undefined;

export function setEmbeddingProviderForTests(provider: EmbeddingProvider | undefined): void {
  overrideProvider = provider;
}

export function getEmbeddingDim(): number {
  const raw = Number(process.env.EMBEDDING_DIM ?? getConfigSafe()?.EMBEDDING_DIM ?? EMBEDDING_DIM);
  return Number.isInteger(raw) && raw > 0 ? raw : EMBEDDING_DIM;
}

/** Server-side Gemini key. STRICT: never the Anthropic key. */
export function getGeminiEmbeddingKey(): string {
  return process.env.GEMINI_API_KEY?.trim() || '';
}

export function getGeminiEmbeddingModel(): string {
  return process.env.EMBEDDING_MODEL?.trim() || 'gemini-embedding-001';
}

export function isGeminiEmbeddingModel(model = process.env.EMBEDDING_MODEL?.trim() || ''): boolean {
  return model.toLowerCase().startsWith('gemini');
}

function getConfigSafe(): { EMBEDDING_DIM?: string } | undefined {
  try {
    return getConfig() as unknown as { EMBEDDING_DIM?: string };
  } catch {
    return undefined;
  }
}

export function deterministicEmbedding(text: string, dim = getEmbeddingDim()): number[] {
  const vec = new Array<number>(dim).fill(0);
  const tokens = text.toLowerCase().split(/[^a-z0-9]+/).filter(Boolean);
  for (const token of tokens) {
    const h = createHash('sha256').update(token).digest();
    const i1 = h.readUInt16BE(0) % dim;
    const i2 = h.readUInt16BE(2) % dim;
    vec[i1] += 1;
    vec[i2] += 0.5;
  }
  const norm = Math.sqrt(vec.reduce((s, v) => s + v * v, 0));
  if (norm > 0) {
    for (let i = 0; i < vec.length; i++) vec[i] /= norm;
  }
  return vec;
}

export function cosineSimilarity(a: number[], b: number[]): number {
  const n = Math.min(a.length, b.length);
  let dot = 0;
  let na = 0;
  let nb = 0;
  for (let i = 0; i < n; i++) {
    dot += a[i] * b[i];
    na += a[i] * a[i];
    nb += b[i] * b[i];
  }
  if (na === 0 || nb === 0) return 0;
  return dot / (Math.sqrt(na) * Math.sqrt(nb));
}

class AnthropicEmbeddingStub implements EmbeddingProvider {
  name = 'anthropic-stub';
  dim = getEmbeddingDim();
  async embed(texts: string[]): Promise<number[][]> {
    logger.warn('Anthropic embeddings not wired over HTTP in core backend; using deterministic fallback (tests/dev only)');
    return texts.map((t) => deterministicEmbedding(t, this.dim));
  }
}

/**
 * Phase 7 real embedding provider (Voyage-compatible HTTP API).
 * - Key only from environment (EMBEDDING_API_KEY, explicit only).
 * - Configurable model via EMBEDDING_MODEL (default voyage-3-lite).
 * - Timeout + one safe retry (embeddings are idempotent/safe to retry).
 * - Dimension validated before returning; mismatches throw (callers must not
 *   insert wrong-dim vectors into pgvector).
 */
class VoyageEmbeddingProvider implements EmbeddingProvider {
  name = 'voyage';
  dim = getEmbeddingDim();
  async embed(texts: string[]): Promise<number[][]> {
    // Live-verification fix: ONLY an explicit embedding credential enables
    // the Voyage HTTP path. An Anthropic LLM key is not valid there — using
    // it as fallback turned graceful deterministic retrieval into a hard
    // Voyage 401. Empty/missing key → deterministic fallback (tests/dev, or
    // LLM-only deployments).
    const apiKey = process.env.EMBEDDING_API_KEY ?? '';
    const model = process.env.EMBEDDING_MODEL || 'voyage-3-lite';
    if (!apiKey) {
      logger.warn('EMBEDDING_API_KEY unset — using deterministic fallback (tests/dev only)');
      return texts.map((t) => deterministicEmbedding(t, this.dim));
    }
    const baseUrl = (process.env.EMBEDDING_BASE_URL?.trim() || 'https://api.voyageai.com').replace(/\/$/, '');
    const timeoutMs = Number(process.env.EMBEDDING_TIMEOUT_MS ?? 30000);
    const expectedDim = this.dim;
    let lastError: unknown = null;
    for (let attempt = 0; attempt < 2; attempt++) {
      try {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), timeoutMs);
        try {
          const res = await fetch(`${baseUrl}/v1/embeddings`, {
            method: 'POST',
            headers: { 'content-type': 'application/json', authorization: `Bearer ${apiKey}` },
            body: JSON.stringify({ model, input: texts }),
            signal: controller.signal,
          });
          if (!res.ok) {
            const text = await res.text().catch(() => '');
            throw new Error(`Embedding provider error: ${res.status} ${text.slice(0, 200)}`);
          }
          const body = (await res.json()) as { data?: { embedding: number[] }[] };
          const vectors = (body.data ?? []).map((d) => d.embedding);
          if (vectors.length !== texts.length) throw new Error('Embedding provider returned mismatched count');
          for (const v of vectors) validateEmbeddingDim(v, expectedDim);
          return vectors;
        } finally {
          clearTimeout(timer);
        }
      } catch (err) {
        lastError = err;
        logger.warn({ err: err instanceof Error ? err.message : String(err), attempt }, 'embedding provider attempt failed');
        if (attempt === 0) continue;
        throw err instanceof Error ? err : new Error(String(err));
      }
    }
    throw lastError instanceof Error ? lastError : new Error('Embedding provider failed');
  }
}

/**
 * Free-tier runtime embedding provider (Google Gemini, gemini-embedding-001).
 * Same EmbeddingProvider interface as the Voyage path. Requests
 * outputDimensionality = configured dim (1536 — pgvector column unchanged)
 * and validates every vector before returning.
 */
class GeminiEmbeddingProvider implements EmbeddingProvider {
  name = 'gemini';
  dim = getEmbeddingDim();
  async embed(texts: string[]): Promise<number[][]> {
    const apiKey = getGeminiEmbeddingKey();
    const model = getGeminiEmbeddingModel();
    if (!apiKey) {
      logger.warn('GEMINI_API_KEY unset — using deterministic fallback (tests/dev only)');
      return texts.map((t) => deterministicEmbedding(t, this.dim));
    }
    const expectedDim = this.dim;
    const timeoutMs = Number(process.env.EMBEDDING_TIMEOUT_MS ?? 30000);
    const out: number[][] = [];
    for (const text of texts) {
      let lastError: unknown = null;
      for (let attempt = 0; attempt < 2; attempt++) {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), timeoutMs);
        try {
          const res = await fetch(
            `https://generativelanguage.googleapis.com/v1beta/models/${encodeURIComponent(model)}:embedContent`,
            {
              method: 'POST',
              headers: { 'content-type': 'application/json', 'x-goog-api-key': apiKey },
              body: JSON.stringify({
                model: `models/${model}`,
                content: { parts: [{ text }] },
                outputDimensionality: expectedDim,
              }),
              signal: controller.signal,
            },
          );
          if (!res.ok) {
            const errText = await res.text().catch(() => '');
            throw new Error(`Embedding provider error: ${res.status} ${errText.slice(0, 200)}`);
          }
          const body = (await res.json()) as { embedding?: { values?: number[] } };
          const vec = (body.embedding?.values ?? []).map(Number);
          validateEmbeddingDim(vec, expectedDim);
          out.push(vec);
          lastError = null;
          break;
        } catch (err) {
          lastError = err;
          logger.warn({ err: err instanceof Error ? err.message : String(err), attempt }, 'embedding provider attempt failed');
          if (attempt === 0) continue;
          throw err instanceof Error ? err : new Error(String(err));
        } finally {
          clearTimeout(timer);
        }
      }
      if (lastError) throw lastError instanceof Error ? lastError : new Error(String(lastError));
    }
    return out;
  }
}

/** Validate vector dimension before pgvector insert (PRD: 1536). */
export function validateEmbeddingDim(vec: number[], expectedDim = getEmbeddingDim()): void {  if (!Array.isArray(vec) || vec.length !== expectedDim) {
    throw new Error(`Embedding dimension mismatch: got ${Array.isArray(vec) ? vec.length : 'invalid'}, expected ${expectedDim}`);
  }
  if (!vec.every((v) => Number.isFinite(v))) {
    throw new Error('Embedding contains non-finite values');
  }
}

export function getEmbeddingProvider(): EmbeddingProvider {
  if (overrideProvider) return overrideProvider;
  // Explicit routing (free-tier runtime = Gemini):
  // - model names gemini-* → Gemini provider (falls back internally w/o key)
  // - explicit Voyage credential → Voyage provider (compat)
  // - Gemini key present → Gemini provider
  // - otherwise deterministic stub (tests / LLM-only deployments)
  if (isGeminiEmbeddingModel()) return new GeminiEmbeddingProvider();
  const voyageKey = process.env.EMBEDDING_API_KEY ?? '';
  if (voyageKey) return new VoyageEmbeddingProvider();
  if (getGeminiEmbeddingKey()) return new GeminiEmbeddingProvider();
  return new AnthropicEmbeddingStub();
}

export function serializeEmbedding(vec: number[]): string {
  return JSON.stringify(vec.map((v) => Number(v.toFixed(6))));
}

export function parseEmbedding(raw: unknown, dim = getEmbeddingDim()): number[] | null {
  if (Array.isArray(raw) && raw.every((v) => typeof v === 'number')) {
    return (raw as number[]).slice(0, dim);
  }
  if (typeof raw === 'string') {
    try {
      const parsed: unknown = JSON.parse(raw);
      if (Array.isArray(parsed)) {
        const nums = (parsed as unknown[]).map(Number).filter((v) => Number.isFinite(v));
        return nums.slice(0, dim);
      }
    } catch {
      return null;
    }
  }
  return null;
}

export interface AiServiceEmbeddingClient {
  embed(texts: string[]): Promise<number[][]>;
  dim: number;
}

export function createAiServiceEmbeddingClient(baseUrl: string, token: string, dim = getEmbeddingDim()): AiServiceEmbeddingClient {
  return {
    dim,
    async embed(texts: string[]): Promise<number[][]> {
      const res = await fetch(`${baseUrl.replace(/\/$/, '')}/internal/embed`, {
        method: 'POST',
        headers: { 'content-type': 'application/json', ...(token ? { 'x-ai-service-token': token } : {}) },
        body: JSON.stringify({ texts }),
      });
      if (!res.ok) throw new Error(`AI service embed failed: ${res.status}`);
      const body = (await res.json()) as { embeddings: number[][] };
      return body.embeddings;
    },
  };
}
